sbi/inference/trainers/base.py

```diff
@@ -306,6 +320,193 @@ def train(
+    def _get_potential_function(
+        self,
+        prior: Distribution,
+        estimator: Union[RatioEstimator, ConditionalEstimator],
+    ) -> Tuple[BasePotential, TorchTransform]:
+        """Subclass-specific potential creation"""
+        pass
+
+    def build_posterior(
+        self,
+        estimator: Optional[Union[RatioEstimator, ConditionalEstimator]],
+        prior: Optional[Distribution],
+        sample_with: Literal[
+            "mcmc", "rejection", "vi", "importance", "direct", "sde", "ode"
+        ],
+        **kwargs,
+    ) -> NeuralPosterior:
+        r"""Method for building posteriors.
+
+        This method serves as a base method for constructing a posterior based
+        on a given estimator and prior. The posterior can be sampled using one of
+        several inference methods specified by `sample_with`.
+
+        Args:
+            estimator: The estimator that the posterior is based on.
+            prior: A probability distribution that expresses prior knowledge about the
+                parameters, e.g. which ranges are meaningful for them. Must be a PyTorch
+                distribution, see FAQ for details on how to use custom distributions.
+            sample_with: The inference method to use. Must be one of:
+                - "mcmc"
+                - "rejection"
+                - "vi"
+                - "importance"
+                - "direct"
+                - "sde"
+                - "ode"
+            **kwargs: Additional method-specific parameters.
+
+        Returns:
+            NeuralPosterior object.
+        """
+
+        if prior is None:
+            cls_name = self.__class__.__name__
+            assert (
+                self._prior is not None
+            ), f"""You did not pass a prior. You have to pass the prior either at
+                initialization `inference = {cls_name}(prior)` or to `.build_posterior
+                (prior=prior)`."""
+            prior = self._prior
+        else:
+            check_prior(prior)
+
+        if estimator is None:
+            estimator = self._neural_net
+            # If internal net is used device is defined.
+            device = self._device
+        else:
+            if not isinstance(estimator, (ConditionalEstimator, RatioEstimator)):
+                raise TypeError(
+                    "estimator must be ConditionalEstimator or RatioEstimator,",
+                    f" got {type(estimator).__name__}",
+                )
+            # Otherwise, infer it from the device of the net parameters.
+            device = str(next(estimator.parameters()).device)
+
+        assert estimator is not None, "Expected a valid estimator object, but got None."
+
+        self._posterior = self._create_posterior(
+            estimator,
+            sample_with,
+            prior,
+            device,
+            **kwargs,
+        )
+
+        # Store models at end of each round.
+        self._model_bank.append(deepcopy(self._posterior))
```

Comment:
This method has very mixed levels of abstraction. The error handling code has no abstraction and therefore takes 90 percent of the space and all the actual work is done with one call to `_create_posterior`. I would prefer this method to read like 

```python
_fail_if_prior_is_invalid(prior, self._prior)
_fail_if_estimator_is_invalid(....)
...
self._posterior = self._create_posterior(...)
```

Maybe the `self._create_posterior` method could even be split up in multiple steps to get a better impression of the steps 
- get the estimator
- get the potential function
- get the posterior

Ideally each of these steps would be a method o function call and if we are lucky then each function call only needs a subset of the arguments of `_create_posterior`.

Reply:
very good points!